pep-0612.rst

```diff
@@ -128,169 +156,346 @@ will defer the standardization of the semantics of those options to a later PEP.
-the following example).
+Previously, the only things acceptable in the first "argument" to the
+``typing.Callable`` type are a list of parameter arguments (``[A, B, C]``) or an
+ellipsis (signifying "undefined parameters").  We now augment that with two new
+options:
+
+.. code-block::
+
+   parameters_expression ::=
+     | "..."
+     | "[" [ type_expression ("," type_expression)\* ] "]"
+     | parameter_specification_variable
+     | concatenate "["
+                      type_expression ("," type_expression)\* ","
+                      parameter_specification_variable
+                   "]"
+
+Where ``parameter_specification_variable`` resolves to a ``typing.ParamSpec``
+declaration as defined above, and ``concatenate`` resolves to
+``typing.type_variable_operators.Concatenate``.
+
+As before, ``parameters_expression`` is the only valid thing for the first
+parameter of ``typing.Callable``, but it is now acceptable in other user-defined
+generic types, as long as they are declared as being generic in
+a parameters_expression.
+
+.. code-block::
+
+   T = TypeVar("T")
+   S = TypeVar("S")
+   P_2 = ParamSpec("P_2")
+
+   class X(Generic[T, P]):
+     ...
+
+   def f(x: X[int, P_2]) -> str: ...                    # Accepted
+   def f(x: X[int, Concatenate[int, P_2]]) -> str: ...  # Accepted
+   def f(x: X[int, [int, bool]]) -> str: ...            # Accepted
+   def f(x: X[int, ...]) -> str: ...                    # Accepted
+
+   class Y(Generic[T, Concatenate[S, P]]):
+     ...
+
+   def f(x: Y[int, P_2]) -> str: ...                    # Accepted
+   def f(x: Y[int, Concatenate[int, P_2]]) -> str: ...  # Accepted
+   def f(x: Y[int, [int, bool]]) -> str: ...            # Accepted
+   def f(x: Y[int, ...]) -> str: ...                    # Accepted
+
+We furthermore accept a special case for when a class is generic with
+respect to only a single ParamSpec.
+
+.. code-block::
+
+   class Z(Generic[P]):
+     ...
+
+   def f(x: X[[int, str, bool]]) -> str: ...   # Accepted
+   def f(x: X[int, str, bool]) -> str: ...     # Also accepted
+
+As before, ``parameters_expression``\ s by themselves are not acceptable in
+places where a type is expected
 
 .. code-block::
 
-   def foo(
-       x: typing.Callable[TParams, int]
-   ) -> typing.Callable[TParams, str]:  # Accepted
-       ...
-   def foo(
-       x: MyClass[TParams, int]
-   ) -> typing.Callable[TParams, str]:  # Accepted
-       ...
-   def foo(x: TParams) -> TParams: ...  # Rejected
-   def foo(x: typing.List[TParams]) -> None: ... # Rejected
-   def foo(x: typing.Callable[[int, str], TParams]) -> None: ... # Rejected
+   def foo(x: P) -> P: ...                           # Rejected
+   def foo(x: typing.List[P]) -> None: ...           # Rejected
+   def foo(x: Callable[[int, str], P]) -> None: ...  # Rejected
+
+   def foo(x: Concatenate[int, P]) -> int: ...       # Rejected
+   def foo(x: [int, str, bool]) -> int: ...          # Rejected
```

Comment:
ditto?

Reply:
here I think your case is stronger since I don't think anyone expects this to work.